### PR TITLE
[platform] support for Texas Instruments cc2652

### PR DESCRIPTION
**Pull request by srickardti**
Adds support for the Texas Instruments CC2652R1 SoC. 

Updates CC26XXware to latest version.

Special thanks to @pkanek and @DuaneEllis-TI

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -12,6 +12,7 @@
 *.suo
 *.trs
 *.user
+*.bak
```
Comment (jwhui) [inline .gitignore]: nit: try to maintain alphabetic order (I realize a few entries violate this rule already).

File: examples/apps/cli/Makefile.am [hunk 2]
```diff
@@ -78,20 +78,36 @@ endif # OPENTHREAD_EXAMPLES_CC2538
...
+CPPFLAGS_COMMON                                                       += \
+    -I$(top_srcdir)/third_party/ti/devices/cc26x2                        \
+    $(NULL)
+
+LDADD_COMMON                                                          +=  \
+    $(top_builddir)/examples/platforms/cc2652/libopenthread-cc2652.a      \
+    $(top_srcdir)/third_party/ti/devices/cc26x2/driverlib/bin/gcc/driverlib.a \
+    $(NULL)
+
+LDFLAGS_COMMON                                                        +=   \
```
Comment (jwhui) [inline examples/apps/cli/Makefile.am]: nit: align `\`

File: examples/apps/ncp/Makefile.am [hunk 3]
```diff
@@ -78,20 +78,36 @@ endif # OPENTHREAD_EXAMPLES_CC2538
...
+CPPFLAGS_COMMON                                                       += \
+    -I$(top_srcdir)/third_party/ti/devices/cc26x2                        \
+    $(NULL)
+
+LDADD_COMMON                                                          +=  \
+    $(top_builddir)/examples/platforms/cc2652/libopenthread-cc2652.a      \
+    $(top_srcdir)/third_party/ti/devices/cc26x2/driverlib/bin/gcc/driverlib.a \
+    $(NULL)
+
+LDFLAGS_COMMON                                                        +=   \
```
Comment (jwhui) [inline examples/apps/ncp/Makefile.am]: nit: align `\`

File: examples/platforms/cc2652/Makefile.am [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+include $(abs_top_nlbuild_autotools_dir)/automake/pre.am
+
+lib_LIBRARIES                       = libopenthread-cc2652.a
```
Comment (jwhui) [inline examples/platforms/cc2652/Makefile.am]: nit: align `=`

File: examples/platforms/cc2652/README.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
+# OpenThread on CC2652 Example
+
+This directory contains example platform drivers for the [Texas
+Instruments CC2652R1][cc2652r1].
+
+The example platform drivers are intended to present the minimal code necessary
+to support OpenThread.  As a result, the example platform drivers do not
+necessarily highlight the platform's full capabilities.  The platform drivers
+were build for the [CC2652R1 LAUNCHXL][cc2652r1-launchxl], usage on other
```
Comment (jwhui) [inline examples/platforms/cc2652/README.md]: typo: `build` -> `built`

File: examples/platforms/cc2652/README.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+boards with a cc2652r1 may require changes to the peripheral drivers.
+
+<!---
+TODO: update documentation with live links
```
Comment (jwhui) [inline examples/platforms/cc2652/README.md]: Still a TODO?

File: examples/platforms/cc2652/README.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+[GNU AutoMake][gnu-automake], [Python][python], and the [ARM gcc
+toolchain][arm-toolchain].
+
+With the exception of the arm toolchain, most of these tools are installed by
+default on modern Posix systems. It is recommended to setup a Linux virtual
+machine for building on a Windows host system. For help setting up VirtualBox
+with Ubuntu, consult this [community help wiki
+article][ubuntu-wiki-virtualbox].
+
+
```
Comment (jwhui) [inline examples/platforms/cc2652/README.md]: nit: extra newline

File: examples/platforms/cc2652/cc2652_startup.c [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * startup code is in 'third_party/ti'
+ */
+#if defined(__GNUC__)
+#include <startup_files/startup_gcc.c>
+#elif defined(__TI_ARM__)
+#include <startup_files/startup_ccs.c>
+#else
+#error "Unknown compiler"
+#endif
+
```
Comment (jwhui) [inline examples/platforms/cc2652/cc2652_startup.c]: nit: extra newline

File: examples/platforms/cc2652/crypto/aes_alt.c [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                    retval = MBEDTLS_ERR_AES_INVALID_KEY_LENGTH);
+
+    sUsedKeys |= (1 << key_idx);
+    ctx->key_idx = key_idx;
+exit:
+    return retval;
+}
+
+/**
+ * \brief          AES-ECB block encryption/decryption
```
Comment (jwhui) [inline examples/platforms/cc2652/crypto/aes_alt.c]: There is a mix of `@` vs. `\` for Doxygen keywords in this PR.  OpenThread typically uses `@`.  Can we unify this PR around `@`?

File: examples/platforms/cc2652/crypto/aes_alt.h [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * \brief          AES-ECB block encryption/decryption
+ *
+ * \param ctx      AES context
+ * \param mode     MBEDTLS_AES_ENCRYPT or MBEDTLS_AES_DECRYPT
+ * \param input    16-byte input block
+ * \param output   16-byte output block
+ *
+ * \return         0 if successful
```
Comment (jwhui) [inline examples/platforms/cc2652/crypto/aes_alt.h]: Can this return anything other than 0?

**Comment by jwhui** [comment]
Great to see support for the CC2652!

File: examples/platforms/cc2652/README.md (hunk 6 above)
Comment (srickardti) [inline examples/platforms/cc2652/README.md]: The LaunchPad link is live and I need to update it. The CC2652R1 product page is not live and won't be until we release the product to market, I'll move the TODO to reflect that this link needs to be updated when the product page is live.

Comment (srickardti) [inline examples/platforms/cc2652/README.md]: The LaunchPad link is live and I need to update it. The CC2652R1 product page is not live and won't be until we release the product to market, I'll move the TODO to reflect that this link needs to be updated when the product page is live.

**Comment by srickardti** [comment]
I'll update the READMEs and Doxygen, thanks for the comments.

File: .gitignore (hunk 1 above)
Comment (srickardti) [inline .gitignore]: This was placed next to the ignore for Emacs backup files.

I'll move it up to be in alphabetical order.

File: examples/Makefile-cc2652 [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+LD                              = arm-none-eabi-ld
+STRIP                           = arm-none-eabi-strip
+NM                              = arm-none-eabi-nm
+RANLIB                          = arm-none-eabi-ranlib
+OBJCOPY                         = arm-none-eabi-objcopy
+
+BuildJobs                      ?= 10
+
+configure_OPTIONS               = \
+    --enable-cli-app=all          \
```
Comment (aeliot) [inline examples/Makefile-cc2652]: nig: Can you align all the `\` to the same column?

File: examples/Makefile-cc2652 [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+# configure-arch <arch>
+#
+# Configure OpenThread for the specified architecture.
+#
+#   arch - The architecture to configure.
+#
+define configure-arch
+$(ECHO) "  CONFIG   $(TargetTuple)..."
+(cd $(BuildPath)/$(TargetTuple) && $(AbsTopSourceDir)/configure \
+INSTALL="$(INSTALL) $(INSTALLFLAGS)" \
```
Comment (aeliot) [inline examples/Makefile-cc2652]: nit: Can you align all the `\`

File: examples/platforms/cc2652/alarm.c [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    return ((rtcVal * 1000) >> 32);
+}
+
+/**
+ * Function documented in platform/alarm-milli.h
+ */
+void otPlatAlarmMilliStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
+{
+    (void)aInstance;
+    sTime0 = aT0;
```
Comment (aeliot) [inline examples/platforms/cc2652/alarm.c]: nit: Align `=`

File: examples/platforms/cc2652/diag.c [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+void otPlatDiagProcess(otInstance *aInstance, int argc, char *argv[], char *aOutput, size_t aOutputMaxLen)
+{
+    // Add more platform specific diagnostics features here.
+    if (argc > 1)
+    {
+        snprintf(aOutput, aOutputMaxLen, "diag feature '%s' is not supported\r\n", argv[0]);
+    }
+
+    (void) argc;
```
Comment (aeliot) [inline examples/platforms/cc2652/diag.c]: `argc` is used above, this line is unneeded.

File: examples/platforms/cc2652/flash.c [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+ */
+otError utilsFlashErasePage(uint32_t aAddress)
+{
+    uint32_t mode;
+    uint32_t fsmRet;
+
+    if (isVoltageLow())
+    {
+        /* Supplied voltage was too low. */
+        return OT_ERROR_FAILED;
```
Comment (aeliot) [inline examples/platforms/cc2652/flash.c]: Only one return per function, you can use one of the error handling functions here instead.

File: examples/platforms/cc2652/flash.c [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+ */
+uint32_t utilsFlashWrite(uint32_t aAddress, uint8_t *aData, uint32_t aSize)
+{
+    uint32_t mode;
+    uint32_t written = 0;
+
+    if (isVoltageLow())
+    {
+        /* Supplied voltage was too low. */
+        return OT_ERROR_FAILED;
```
Comment (aeliot) [inline examples/platforms/cc2652/flash.c]: There should only be one return per function

Comment (aeliot) [inline examples/platforms/cc2652/flash.c]: Also, the return type is `uint32_t`, not `otError`

File: examples/platforms/cc2652/radio.c [hunk 17]
```diff
@@ -0,0 +1,1861 @@
...
+ * of the receive queue.
+ */
+static void rfCoreInitBufs(void)
+{
+    rfc_dataEntry_t *entry;
+    memset(sRxBuf0, 0x00, RX_BUF_SIZE);
+    memset(sRxBuf1, 0x00, RX_BUF_SIZE);
+
+    entry = (rfc_dataEntry_t *)sRxBuf0;
+    entry->pNextEntry = sRxBuf1;
```
Comment (aeliot) [inline examples/platforms/cc2652/radio.c]: nit: Align the `=`

File: examples/platforms/cc2652/radio.c [hunk 18]
```diff
@@ -0,0 +1,1861 @@
...
+ */
+static uint8_t rfCoreFindShortSrcMatchIdx(const uint16_t aAddress)
+{
+    uint8_t i;
+
+    for (i = 0; i < CC2652_SHORTADD_SRC_MATCH_NUM; i++)
+    {
+        if (sSrcMatchShortData.extAddrEnt[i].shortAddr == aAddress)
+        {
+            return i;
```
Comment (aeliot) [inline examples/platforms/cc2652/radio.c]: Only one return per function

File: examples/platforms/cc2652/radio.c [hunk 19]
```diff
@@ -0,0 +1,1861 @@
...
+ */
+static uint8_t rfCoreFindEmptyShortSrcMatchIdx(void)
+{
+    uint8_t i;
+
+    for (i = 0; i < CC2652_SHORTADD_SRC_MATCH_NUM; i++)
+    {
+        if ((sSrcMatchShortData.srcMatchEn[i / 32] & (1 << (i % 32))) == 0u)
+        {
+            return i;
```
Comment (aeliot) [inline examples/platforms/cc2652/radio.c]: Only one return per function

File: examples/platforms/cc2652/radio.c [hunk 20]
```diff
@@ -0,0 +1,1861 @@
...
+ * be handled by the @ref RFCCPE0IntHandler interrupt handler.
+ */
+static void rfCoreSetupInt(void)
+{
+    bool interruptsWereDisabled;
+
+    /* We are already turned on by the caller, so this should not happen */
+    if (!PRCMRfReady())
+    {
+        return;
```
Comment (aeliot) [inline examples/platforms/cc2652/radio.c]: The return should be at the end, you can use the error handling functions to then jump to it

File: examples/platforms/cc2652/radio.c [hunk 21]
```diff
@@ -0,0 +1,1861 @@
...
+    if (sState == cc2652_stateSleep)
+    {
+        error = OT_ERROR_NONE;
+    }
+    else if (sState == cc2652_stateReceive)
+    {
+        if (rfCoreExecuteAbortCmd() != CMDSTA_Done)
+        {
+            error = OT_ERROR_BUSY;
+            return error;
```
Comment (aeliot) [inline examples/platforms/cc2652/radio.c]: Only one return per function.

**Comment by srickardti** [comment]
Wasn't really sure about the styling on the makefiles. I aligned the equals sign on the multiple line files at column 60 and aligned the `\` characters with the farthest right needed.

File: examples/platforms/cc2652/misc.c [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+ * Function documented in platform/misc.h
+ */
+otPlatResetReason otPlatGetResetReason(otInstance *aInstance)
+{
+    (void)aInstance;
+
+    switch (SysCtrlResetSourceGet())
+    {
+    case RSTSRC_PWR_ON:
+        return OT_PLAT_RESET_REASON_POWER_ON;
```
Comment (aeliot) [inline examples/platforms/cc2652/misc.c]: One return per function.

File: examples/platforms/cc2652/radio.c [hunk 23]
```diff
@@ -0,0 +1,1875 @@
...
+    interruptsWereDisabled = IntMasterDisable();
+
+    if ((ret = (RFCDoorbellSendTo((uint32_t)&sStartRatCmd) & 0xFF)) != CMDSTA_Done)
+    {
+        if (!interruptsWereDisabled)
+        {
+            IntMasterEnable();
+        }
+
+        return ret;
```
Comment (aeliot) [inline examples/platforms/cc2652/radio.c]: One return per function.

File: examples/platforms/cc2652/radio.c [hunk 24]
```diff
@@ -0,0 +1,1875 @@
...
+    doorbellRet = (RFCDoorbellSendTo((uint32_t)&sFsPowerdownCmd) & 0xFF);
+
+    if (doorbellRet != CMDSTA_Done)
+    {
+        if (!interruptsWereDisabled)
+        {
+            IntMasterEnable();
+        }
+
+        return doorbellRet;
```
Comment (aeliot) [inline examples/platforms/cc2652/radio.c]: One return per function

**Pull request merged**